keps/sig-storage/661-statefulset-volume-resize/README.md

```diff
@@ -0,0 +1,1112 @@
+  * Even though events might not be realiable for gathering feedback (especially during an automated deployment process, eg a CI/CD pipeline), the user could still rely on them for troubleshooting - but in that case, the user could just get/subscribe (`kubectl get events -w`) to events from the entire namespace.
+    * It should also be possible to create tooling that will allow the user to aggregate events from multiple PVCs that were created by the same StatefulSet/template.
+* It will be inconsistent with how events are reported on the StatefulSet pod.
+  * Error events that are generated by kubelet on a pod (eg `Failed to pull image`) are not propagated to the pod's StatefulSet object.
+
+### Feedback
+
+![](./diag2.png)
+
+It is useful to provide feedback to the user about the status of the resize operation. This main benefits are
+
+* For errors that we can't properly report via an event on the StatefulSet object - a delay in feedback could prompt the user to investigate the issue, instead of leaving them oblivious to it. 
+* Even for scenarios where there are no errors, it could be useful to provide feedback to the user that the resize operation has completed or is still in progress. If the user has multiple StatefulSets and/or cluster they are resizing, this will give them the option to decide if they want to wait for the resize operation to complete before proceeding with other operations.
+* Feedback is especially useful if the resize happens as part of an automated deployment mechanism (ie a CI/CD pipeline). In such cases, the user doesn't necessarily keep an eye on the deployment process so proper feedback is really important.  
+
+We propose to provide feedback by adding a new field to the StatefulSet status object. The new field will be an array that contains
+* The name of the volume claim template
+* The current number of replicas for which the size of that PVC template is fully reconciled with the capacity of the PVC object (the `.status.capacity.storage` field) 
+* The last generation (pertains to the `.metadata.generation` field on the StatefulSet object) in which the size of that PVC template was fully reconciled with the capacity of the PVC object for all replicas.  
+
+Here's an example of what the StatefulSet status object with the new field will look like
+
+```yaml
+status:
+  availableReplicas: 3
+  collisionCount: 0
+  currentReplicas: 3
+  currentRevision: ex1-54c5bd476c
+  observedGeneration: 3
+  readyReplicas: 3
+  replicas: 3
+  updateRevision: ex1-54c5bd476c
+  updatedReplicas: 3
+  volumeClaimTemplates:
+  - finishedReconciliationGeneration: 2
+    readyReplicas: 1
+    templateName: vol1
+  - finishedReconciliationGeneration: 3
+    readyReplicas: 3
+    templateName: vol2
+```
+
+Some notes about the new fields
+* The `finishedReconciliationGeneration` field is optional and won't appear if the PVC template is newly created and isn't reconciled yet.
```

Comment:
Not sure I understand. My understanding is that `observedGeneration` is the last `generation` of the statefulset that the statefulset has observed. This can't be used as a reliable feedback mechanism, because it wouldn't give the user feedback that the volume has finished expanding. 
`finishedReconciliationGeneration` on the other hand will be incremented to the latest generation only once the statefulset controller has observed that the PVC has been fully expanded to the requested size

Reply:
`osbserrvedGeneration` is in every object, so I'm proposing to use that instead of adding a new one. But your understanding of the mechanism in case of StatefulSet is correct. I'm proposing to use the exact same mechanism (that users are already familiar with) also for volumes, does that make sense?

Reply:
So you're suggesting to repurpose observedGeneration? Ie modify the statefulset controller to only update it when PVC reconciliation finishes? It's possible and one more benefit is that existing clients won't have to be modified to get feedback on volume reconciliation 
but two things that are worth mentioning
1. With the new field, we get per volume granularity (imagine a statefulset with two volume templates) 
2. Some clients might be interested in getting feedback for when the pods of the statefulset get fully reconciled and wouldn't care about the volume. With the new field, there's better granularity and clients can choose what kind of feedback they need

Reply:
I might be wrong, but I think you can achieve the exact same results with `observedGeneration` like I mentioned before.